File: schemas/file-management-document-storage/drizzle/file_comments.ts
// file_comments: Threaded comments on files — supports nested replies via parent_id self-reference and resolved state for review workflows.
// See README.md for full design rationale.

import {
  pgTable,
  uuid,
  text,
  boolean,
  timestamp,
  index,
  AnyPgColumn,
} from "drizzle-orm/pg-core";
import { files } from "./files";
import { users } from "../../auth-rbac/drizzle/users";

export const fileComments = pgTable(
  "file_comments",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    fileId: uuid("file_id")
      .notNull()
      .references(() => files.id, { onDelete: "cascade" }),
    parentId: uuid("parent_id").references(
      (): AnyPgColumn => fileComments.id,
      { onDelete: "cascade" }
    ), // Parent comment for threaded replies. Null = top-level comment.
    authorId: uuid("author_id")
      .notNull()
      .references(() => users.id, { onDelete: "restrict" }),
    body: text("body").notNull(), // Comment text. Supports plain text or markdown.
    isResolved: boolean("is_resolved").notNull().default(false), // Whether this comment thread is resolved.
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdateFn(() => new Date()),
  },
  (table) => [
    index("idx_file_comments_file_id_created_at").on(table.fileId, table.createdAt),
    index("idx_file_comments_parent_id").on(table.parentId),
    index("idx_file_comments_author_id").on(table.authorId),
  ]
);

File: schemas/file-management-document-storage/spacetimedb/file_comments.rs
// file_comments: Threaded comments on files — supports nested replies via parent_id self-reference and resolved state for review workflows.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = file_comments, public)]
pub struct FileComment {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub file_id: String, // UUID — FK → files.id (cascade delete)

    #[index(btree)]
    pub parent_id: Option<String>, // UUID — FK → file_comments.id (cascade delete). Null = top-level comment.

    #[index(btree)]
    pub author_id: String, // UUID — FK → users.id (restrict delete)

    pub body: String, // Comment text. Supports plain text or markdown.

    pub is_resolved: bool, // Whether this comment thread is resolved. Default false in app logic.

    pub created_at: Timestamp,
    pub updated_at: Timestamp, // Comments can be edited.
}
